### PR TITLE
[FLINK-3444] env.fromElements relies on the first input element for determining the DataSet/DataStream type

### PR DESCRIPTION
**Pull request by gallenvara**
Add `fromElements` method with based class type to avoid the exception. The based class should equal to or be the superclass of classes of other elements.

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 1]
```diff
@@ -673,6 +673,43 @@ public TimeCharacteristic getStreamTimeCharacteristic() {
...
+		if (data.length == 0) {
+			throw new IllegalArgumentException("fromElements needs at least one element as argument");
+		}
+
+		TypeInformation<OUT> typeInfo;
+		try {
+			typeInfo = TypeExtractor.getForClass(clazz);
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Could not create TypeInformation for type " + data[0].getClass().getName()
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: shouldn't this exception contain clazz.getName() instead?

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 2]
```diff
@@ -780,6 +780,33 @@ public CsvReader readCsvFile(String filePath) {
...
+	@SafeVarargs
+	public final <X> DataSource<X> fromElements(Class<X> type, X... data) {
+		if (data == null) {
+			throw new IllegalArgumentException("The data must not be null.");
+		}
+		if (data.length == 0) {
+			throw new IllegalArgumentException("The number of elements must not be zero.");
+		}
+		
+		return fromCollection(Arrays.asList(data), TypeExtractor.getForClass(type), Utils.getCallLocationName());
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: why does the scala variant throw a specific exception on getForClass() failure but this one doesn't?

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 3]
```diff
@@ -780,6 +780,33 @@ public CsvReader readCsvFile(String filePath) {
 		return fromCollection(Arrays.asList(data), TypeExtractor.getForObject(data[0]), Utils.getCallLocationName());
 	}
 	
+	/**
+	 * Creates a new data set that contains the given elements. The elements must all be of the same type,
+	 * for example, all of the {@link String} or {@link Integer}. The sequence of elements must not be empty.
+	 * <p>
+	 * The framework will try and determine the exact type from the collection elements.
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: this line doesn't some valid for this method.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/SourceFunctionTest.java [hunk 4]
```diff
@@ -44,6 +45,18 @@ public void fromElementsTest() throws Exception {
 	}
 
 	@Test
+	public void fromElementsWithBaseTypeTest1() {
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/SourceFunctionTest.java]: these tests appear to be in the wrong file;I would put them into the StreamExecutionEnvironment class.

**Comment by zentol** [comment]
Should this method also be added to the scala environment?

**Comment by gallenvara** [comment]
@zentol Thanks a lot for review work. I will modify the codes base on your advice!

**Comment by gallenvara** [comment]
@zentol , PR updated. The scala environment determine the type with `implicitly[TypeInformation[T]]` which is always the class `Object`. In the case this issue mentioned, no exception is thrown. IMO, we should not add this method to it.

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 5]
```diff
@@ -777,7 +777,50 @@ public CsvReader readCsvFile(String filePath) {
...
 		
-		return fromCollection(Arrays.asList(data), TypeExtractor.getForObject(data[0]), Utils.getCallLocationName());
+		TypeInformation<X> typeInfo;
+		try {
+			typeInfo = TypeExtractor.getForObject(data[0]);
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Could not create TypeInformation for type " + data[0].getClass().getName()
+					+ "; please specify the TypeInformation manually via "
+					+ "StreamExecutionEnvironment#fromCollection(Collection, TypeInformation)");
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: This should be ExecutionEnvironment#fromCollection

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 6]
```diff
@@ -777,7 +777,50 @@ public CsvReader readCsvFile(String filePath) {
...
+		}
+		
+		TypeInformation<X> typeInfo;
+		try {
+			typeInfo = TypeExtractor.getForClass(type);
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Could not create TypeInformation for type " + type.getName()
+					+ "; please specify the TypeInformation manually via "
+					+ "StreamExecutionEnvironment#fromCollection(Collection, TypeInformation)");
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: This should be ExecutionEnvironment#fromCollection

**Comment by gallenvara** [comment]
@zentol codes modified and rebase the new commit with previous one.

**Comment by zentol** [comment]
+1

**Comment by StephanEwen** [comment]
Merging this...

**Comment by gallenvara** [comment]
The error of CI build failure is not relevant with this PR.

**Pull request closed without merging**